code/modules/admin/verbs/maprotation.dm

```diff
@@ -33,12 +33,69 @@
 			mapname += "\]"
 
 		maprotatechoices[mapname] = VM
-	var/chosenmap = input("Choose a map to change to", "Change Map")  as null|anything in sortList(maprotatechoices)
+	var/chosenmap = input("Choose a map to change to", "Change Map")  as null|anything in sortList(maprotatechoices)|"Custom"
 	if (!chosenmap)
 		return
+	
 	SSticker.maprotatechecked = 1
-	var/datum/map_config/VM = maprotatechoices[chosenmap]
-	message_admins("[key_name_admin(usr)] is changing the map to [VM.map_name]")
-	log_admin("[key_name(usr)] is changing the map to [VM.map_name]")
-	if (SSmapping.changemap(VM) == 0)
-		message_admins("[key_name_admin(usr)] has changed the map to [VM.map_name]")
+	if(chosenmap == "Custom")
+		message_admins("[key_name_admin(usr)] is changing the map to a custom map")
+		log_admin("[key_name(usr)] is changing the map to a custom map")
+		var/datum/map_config/VM = new
+
+		VM.map_name = input("Choose the name for the map", "Map Name") as null|text
+		if(isnull(VM.map_name))
+			VM.map_name = "Custom"
+		
+		var/map_file = input("Pick file:", "Map File") as null|file
+		if(isnull(map_file))
+			return
+		
+		if(copytext("[map_file]",-4) != ".dmm")
+			to_chat(src, "<span class='warning'>Filename must end in '.dmm': [map_file]</span>")
+			return
+
+		if(!fcopy(map_file, "_maps/custom/[map_file]"))
+			return
+		
+		// This is to make sure the map works so the server does not start without a map.
+		var/datum/parsed_map/M = new (map_file)
```

Comment:
You should check if parsed bounds are set, also might want to call the general validation

Reply:
That should do it.